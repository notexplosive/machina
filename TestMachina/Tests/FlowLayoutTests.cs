using ApprovalTests;
using ApprovalTests.Reporters;
using FluentAssertions;
using Machina.Data;
using Machina.Data.Layout;
using Microsoft.Xna.Framework;
using TestMachina.Utility;
using Xunit;

namespace TestMachina.Tests
{
    [UseReporter(typeof(DiffReporter))]
    public class FlowLayoutTests
    {
        [Fact]
        public void can_create_flow_layout()
        {
            var layout = FlowLayout.HorizontalFlowParent("root", LayoutSize.Pixels(30, 30), FlowLayoutStyle.Empty,
                LayoutNode.Leaf("itemA", LayoutSize.Pixels(10, 10)),
                LayoutNode.Leaf("itemB", LayoutSize.Pixels(10, 10)),
                LayoutNode.Leaf("itemC", LayoutSize.Pixels(10, 10)),
                LayoutNode.Leaf("itemD", LayoutSize.Pixels(10, 10)),
                LayoutNode.Leaf("itemE", LayoutSize.Pixels(10, 10)),
                LayoutNode.Leaf("itemF", LayoutSize.Pixels(10, 10))
            );

            var result = layout.Bake();
            Approvals.Verify(LayoutNodeUtils.DrawResult(result));
        }

        [Fact]
        public void flow_layout_can_have_margin()
        {
            var layout = FlowLayout.HorizontalFlowParent("root", LayoutSize.Pixels(40, 40), new FlowLayoutStyle(margin: new Point(5)),
                LayoutNode.Leaf("itemA", LayoutSize.Pixels(10, 10)),
                LayoutNode.Leaf("itemB", LayoutSize.Pixels(10, 10)),
                LayoutNode.Leaf("itemC", LayoutSize.Pixels(10, 10)),
                LayoutNode.Leaf("itemD", LayoutSize.Pixels(10, 10)),
                LayoutNode.Leaf("itemE", LayoutSize.Pixels(10, 10)),
                LayoutNode.Leaf("itemF", LayoutSize.Pixels(10, 10))
            );

            var result = layout.Bake();
            Approvals.Verify(LayoutNodeUtils.DrawResult(result));
        }

        [Fact]
        public void flow_layout_can_have_padding_between_rows()
        {
            var layout = FlowLayout.HorizontalFlowParent("root", LayoutSize.Pixels(30, 30), new FlowLayoutStyle(paddingBetweenRows: 3),
                LayoutNode.Leaf("itemA", LayoutSize.Pixels(10, 10)),
                LayoutNode.Leaf("itemB", LayoutSize.Pixels(10, 10)),
                LayoutNode.Leaf("itemC", LayoutSize.Pixels(10, 10)),
                LayoutNode.Leaf("itemD", LayoutSize.Pixels(10, 10)),
                LayoutNode.Leaf("itemE", LayoutSize.Pixels(10, 10)),
                LayoutNode.Leaf("itemF", LayoutSize.Pixels(10, 10))
            );

            var result = layout.Bake();
            Approvals.Verify(LayoutNodeUtils.DrawResult(result));
        }

        [Fact]
        public void flow_layout_supports_vertical_alignment()
        {
            var layout = FlowLayout.HorizontalFlowParent("root", LayoutSize.Pixels(30, 30), new FlowLayoutStyle(alignment: Alignment.BottomLeft),
                LayoutNode.Leaf("itemA", LayoutSize.Pixels(10, 10)),
                LayoutNode.Leaf("itemB", LayoutSize.Pixels(10, 10)),
                LayoutNode.Leaf("itemC", LayoutSize.Pixels(10, 10)),
                LayoutNode.Leaf("itemD", LayoutSize.Pixels(10, 10)),
                LayoutNode.Leaf("itemE", LayoutSize.Pixels(10, 10)),
                LayoutNode.Leaf("itemF", LayoutSize.Pixels(10, 10))
            );

            var result = layout.Bake();
            Approvals.Verify(LayoutNodeUtils.DrawResult(result));
        }


        [Fact]
        public void flow_layout_supports_horizontal_alignment()
        {
            var layout = FlowLayout.HorizontalFlowParent("root", LayoutSize.Pixels(40, 30), new FlowLayoutStyle(alignment: Alignment.CenterRight),
                LayoutNode.Leaf("itemA", LayoutSize.Pixels(12, 10)),
                LayoutNode.Leaf("itemB", LayoutSize.Pixels(7, 10)),
                LayoutNode.Leaf("itemC", LayoutSize.Pixels(7, 10)),
                LayoutNode.Leaf("itemD", LayoutSize.Pixels(7, 10)),
                LayoutNode.Leaf("itemE", LayoutSize.Pixels(7, 10)),
                LayoutNode.Leaf("itemF", LayoutSize.Pixels(7, 10))
            );

            var result = layout.Bake();
            Approvals.Verify(LayoutNodeUtils.DrawResult(result));
        }

        [Fact]
        public void flow_layout_supports_horizontal_alignment_center()
        {
            var layout = FlowLayout.HorizontalFlowParent("root", LayoutSize.Pixels(40, 30), new FlowLayoutStyle(alignment: Alignment.Center),
                LayoutNode.Leaf("itemA", LayoutSize.Pixels(12, 10)),
                LayoutNode.Leaf("itemB", LayoutSize.Pixels(7, 10)),
                LayoutNode.Leaf("itemC", LayoutSize.Pixels(9, 10)),
                LayoutNode.Leaf("itemD", LayoutSize.Pixels(13, 10)),
                LayoutNode.Leaf("itemE", LayoutSize.Pixels(7, 10)),
                LayoutNode.Leaf("itemF", LayoutSize.Pixels(7, 10))
            );

            var result = layout.Bake();
            Approvals.Verify(LayoutNodeUtils.DrawResult(result));
        }

        [Fact]
        public void flow_layout_supports_padding_between_individual_items()
        {
            var layout = FlowLayout.HorizontalFlowParent("root", LayoutSize.Pixels(40, 30), new FlowLayoutStyle(paddingBetweenItemsInEachRow: 4),
                LayoutNode.Leaf("itemA", LayoutSize.Pixels(12, 10)),
                LayoutNode.Leaf("itemB", LayoutSize.Pixels(7, 10)),
                LayoutNode.Leaf("itemC", LayoutSize.Pixels(9, 10)),
                LayoutNode.Leaf("itemD", LayoutSize.Pixels(13, 10)),
                LayoutNode.Leaf("itemE", LayoutSize.Pixels(7, 10)),
                LayoutNode.Leaf("itemF", LayoutSize.Pixels(7, 10))
            );

            var result = layout.Bake();
            Approvals.Verify(LayoutNodeUtils.DrawResult(result));
        }

        [Fact]
        public void flow_layout_estimates_accurate_height()
        {
            var layout = FlowLayout.HorizontalFlowParent("root", LayoutSize.Pixels(40, 30), FlowLayoutStyle.Empty,
                LayoutNode.Leaf("itemA", LayoutSize.Pixels(12, 5)),
                LayoutNode.Leaf("itemB", LayoutSize.Pixels(13, 7)),
                LayoutNode.Leaf("itemC", LayoutSize.Pixels(12, 3)),
                LayoutNode.Leaf("itemD", LayoutSize.Pixels(13, 12)),
                LayoutNode.Leaf("itemE", LayoutSize.Pixels(8, 15)),
                LayoutNode.Leaf("itemF", LayoutSize.Pixels(10, 5))
            );

            var result = layout.Bake();
            Approvals.Verify(LayoutNodeUtils.DrawResult(result));
        }

        [Fact]
        public void flow_layout_can_permit_overflow_extra_rows()
        {
            var layout = FlowLayout.HorizontalFlowParent("root", LayoutSize.Pixels(40, 20), new FlowLayoutStyle(overflowRule: OverflowRule.Free),
                LayoutNode.Leaf("itemA", LayoutSize.Pixels(12, 10)),
                LayoutNode.Leaf("itemB", LayoutSize.Pixels(7, 10)),
                LayoutNode.Leaf("itemC", LayoutSize.Pixels(9, 10)),
                LayoutNode.Leaf("itemD", LayoutSize.Pixels(13, 10)),
                LayoutNode.Leaf("itemE", LayoutSize.Pixels(7, 10)),
                LayoutNode.Leaf("itemF", LayoutSize.Pixels(8, 10)),
                LayoutNode.Leaf("itemG", LayoutSize.Pixels(9, 10)),
                LayoutNode.Leaf("itemH", LayoutSize.Pixels(4, 4))
            );

            var result = layout.Bake();
            Approvals.Verify("There are supposed to be a bunch of skipped pixels here: \n\n" + LayoutNodeUtils.DrawResult(result));
        }

        [Fact]
        public void flow_layout_halts_on_illegal_overflow()
        {
            var layout = FlowLayout.HorizontalFlowParent("root", LayoutSize.Pixels(40, 20), new FlowLayoutStyle(overflowRule: OverflowRule.EverythingMustBeInside),
                LayoutNode.Leaf("itemA", LayoutSize.Pixels(12, 10)),
                LayoutNode.Leaf("itemB", LayoutSize.Pixels(7, 10)),
                LayoutNode.Leaf("itemC", LayoutSize.Pixels(9, 10)),
                LayoutNode.Leaf("itemD", LayoutSize.Pixels(13, 10)),
                LayoutNode.Leaf("itemE", LayoutSize.Pixels(7, 10)),
                LayoutNode.Leaf("itemF", LayoutSize.Pixels(8, 12)),
                LayoutNode.Leaf("itemG", LayoutSize.Pixels(9, 10))
            );

            var result = layout.Bake();
            Approvals.Verify(LayoutNodeUtils.DrawResult(result));
        }

        [Fact]
        public void flow_layout_halts_on_illegal_but_keep_last_overflow()
        {
            var layout = FlowLayout.HorizontalFlowParent("root", LayoutSize.Pixels(40, 20), new FlowLayoutStyle(overflowRule: OverflowRule.LastRowKeepsGoing),
                LayoutNode.Leaf("itemA", LayoutSize.Pixels(12, 10)),
                LayoutNode.Leaf("itemB", LayoutSize.Pixels(7, 10)),
                LayoutNode.Leaf("itemC", LayoutSize.Pixels(9, 10)),
                LayoutNode.Leaf("itemD", LayoutSize.Pixels(13, 10)),
                LayoutNode.Leaf("itemE", LayoutSize.Pixels(7, 10)),
                LayoutNode.Leaf("really-wide-item", LayoutSize.Pixels(21, 10))
            );

            var result = layout.Bake();
            Approvals.Verify(LayoutNodeUtils.DrawResult(result));
        }

        [Fact]
        public void flow_layout_allows_forced_linebreaks()
        {
            var layout = FlowLayout.HorizontalFlowParent("root", LayoutSize.Pixels(40, 40), FlowLayoutStyle.Empty,
                LayoutNode.Leaf("itemA", LayoutSize.Pixels(12, 10)),
                LayoutNode.Leaf("itemB", LayoutSize.Pixels(7, 10)),
                FlowLayoutInstruction.Linebreak,
                LayoutNode.Leaf("itemC", LayoutSize.Pixels(9, 5)),
                LayoutNode.Leaf("itemD", LayoutSize.Pixels(13, 7)),
                LayoutNode.Leaf("itemE", LayoutSize.Pixels(7, 5)),
                FlowLayoutInstruction.Linebreak,
                LayoutNode.Leaf("itemF", LayoutSize.Pixels(8, 10)),
                LayoutNode.Leaf("itemG", LayoutSize.Pixels(9, 10))
            );

            var result = layout.Bake();
            Approvals.Verify(LayoutNodeUtils.DrawResult(result));
        }

        [Fact]
        public void flow_layout_can_be_vertical()
        {
            var layout = FlowLayout.VerticalFlowParent("root", LayoutSize.Pixels(25, 40), new FlowLayoutStyle(paddingBetweenItemsInEachRow: 4),
                LayoutNode.Leaf("itemA", LayoutSize.Pixels(12, 10)),
                LayoutNode.Leaf("itemB", LayoutSize.Pixels(7, 10)),
                LayoutNode.Leaf("itemC", LayoutSize.Pixels(9, 10)),
                LayoutNode.Leaf("itemD", LayoutSize.Pixels(10, 10)),
                LayoutNode.Leaf("itemE", LayoutSize.Pixels(13, 10)),
                LayoutNode.Leaf("itemF", LayoutSize.Pixels(7, 10))
            );

            var result = layout.Bake();
            Approvals.Verify(LayoutNodeUtils.DrawResult(result));
        }

        [Fact]
        public void flow_layout_can_be_vertical_with_alignment()
        {
            var layout = FlowLayout.VerticalFlowParent("root", LayoutSize.Pixels(25, 40), new FlowLayoutStyle(alignment: Alignment.BottomRight),
                LayoutNode.Leaf("itemA", LayoutSize.Pixels(12, 10)),
                LayoutNode.Leaf("itemB", LayoutSize.Pixels(7, 10)),
                LayoutNode.Leaf("itemC", LayoutSize.Pixels(9, 10)),
                LayoutNode.Leaf("itemD", LayoutSize.Pixels(10, 10)),
                LayoutNode.Leaf("itemE", LayoutSize.Pixels(13, 10)),
                LayoutNode.Leaf("itemF", LayoutSize.Pixels(7, 10))
            );

            var result = layout.Bake();
            Approvals.Verify(LayoutNodeUtils.DrawResult(result));
        }

        [Fact]
        public void flow_layout_can_address_individual_rows()
        {
            var layout = FlowLayout.HorizontalFlowParent("root", LayoutSize.Pixels(25, 40), new FlowLayoutStyle(alignment: Alignment.Center),
                LayoutNode.Leaf("itemA", LayoutSize.Pixels(12, 10)),
                LayoutNode.Leaf("itemB", LayoutSize.Pixels(7, 10)),
                FlowLayoutInstruction.Linebreak,
                LayoutNode.Leaf("itemC", LayoutSize.Pixels(9, 10)),
                FlowLayoutInstruction.Linebreak,
                LayoutNode.Leaf("itemD", LayoutSize.Pixels(10, 10)),
                LayoutNode.Leaf("itemE", LayoutSize.Pixels(13, 10)),
                LayoutNode.Leaf("itemF", LayoutSize.Pixels(7, 10)),
                LayoutNode.Leaf("itemG", LayoutSize.Pixels(7, 10))
            );

            var result = layout.Bake();

            result.GetRow(0).ItemCount.Should().Be(2);
            result.GetRow(1).ItemCount.Should().Be(1);
            result.GetRow(2).ItemCount.Should().Be(2);
            result.GetRow(3).ItemCount.Should().Be(2);

            result.GetRow(2).GetItemNode(1).Size.Should().Be(new Point(13, 10));

            result.GetRow(0).Node.Rectangle.Should().Be(new Rectangle(new Point(0, 0), new Point(25, 10)));
            result.GetRow(1).Node.Rectangle.Should().Be(new Rectangle(new Point(0, 10), new Point(25, 10)));
            result.GetRow(2).Node.Rectangle.Should().Be(new Rectangle(new Point(0, 20), new Point(25, 10)));
            result.GetRow(3).Node.Rectangle.Should().Be(new Rectangle(new Point(0, 30), new Point(25, 10)));

            result.GetRow(0).UsedRectangle.Should().Be(new Rectangle(new Point(3, 0), new Point(19, 10)));
            result.GetRow(1).UsedRectangle.Should().Be(new Rectangle(new Point(8, 10), new Point(9, 10)));
            result.GetRow(2).UsedRectangle.Should().Be(new Rectangle(new Point(1, 20), new Point(23, 10)));
            result.GetRow(3).UsedRectangle.Should().Be(new Rectangle(new Point(5, 30), new Point(14, 10)));


            int totalItems = 0;
            foreach (var row in result.Rows)
            {
                foreach (var item in row)
                {
                    item.NestingLevel.Should().Be(3);
                    totalItems++;
                }
            }

            totalItems.Should().Be(7);
        }

        [Fact]
        public void verify_used_rows_horizontal()
        {
            var layout = FlowLayout.HorizontalFlowParent("root", LayoutSize.Pixels(25, 20), new FlowLayoutStyle(alignment: Alignment.Center),
                LayoutNode.Leaf("itemA", LayoutSize.Pixels(12, 5)),
                LayoutNode.Leaf("itemB", LayoutSize.Pixels(7, 5)),
                FlowLayoutInstruction.Linebreak,
                LayoutNode.Leaf("itemC", LayoutSize.Pixels(9, 5)),
                FlowLayoutInstruction.Linebreak,
                LayoutNode.Leaf("itemD", LayoutSize.Pixels(10, 5)),
                LayoutNode.Leaf("itemE", LayoutSize.Pixels(13, 5)),
                LayoutNode.Leaf("itemF", LayoutSize.Pixels(7, 5)),
                LayoutNode.Leaf("itemG", LayoutSize.Pixels(7, 5))
            );

            var result = layout.Bake();

            Approvals.Verify(
                $"Layout\n{LayoutNodeUtils.DrawResult(result)}\n\nUsed Row Rectangles:\n{LayoutNodeUtils.DrawUsedRectangles(result, result.Rows)}\n\nJust Items:\n{LayoutNodeUtils.DrawItems(result, result.Rows)}"
            );
        }

        [Fact]
        public void verify_used_rows_vertical()
        {
            var layout = FlowLayout.VerticalFlowParent("root", LayoutSize.Pixels(40, 30), new FlowLayoutStyle(alignment: Alignment.Center, paddingBetweenItemsInEachRow: 2, paddingBetweenRows: 2),
                LayoutNode.Leaf("itemA", LayoutSize.Pixels(12, 5)),
                LayoutNode.Leaf("itemB", LayoutSize.Pixels(7, 5)),
                FlowLayoutInstruction.Linebreak,
                LayoutNode.Leaf("itemC", LayoutSize.Pixels(9, 5)),
                FlowLayoutInstruction.Linebreak,
                LayoutNode.Leaf("itemD", LayoutSize.Pixels(10, 5)),
                LayoutNode.Leaf("itemE", LayoutSize.Pixels(13, 5)),
                LayoutNode.Leaf("itemF", LayoutSize.Pixels(7, 5)),
                LayoutNode.Leaf("itemG", LayoutSize.Pixels(7, 5))
            );

            var result = layout.Bake();

            Approvals.Verify(
                $"Layout\n{LayoutNodeUtils.DrawResult(result)}\n\nUsed Row Rectangles:\n{LayoutNodeUtils.DrawUsedRectangles(result, result.Rows)}\n\nJust Items:\n{LayoutNodeUtils.DrawItems(result, result.Rows)}"
            );
        }
    }
}