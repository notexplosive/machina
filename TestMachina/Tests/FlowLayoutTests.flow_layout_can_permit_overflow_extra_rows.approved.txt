There are supposed to be a bunch of skipped pixels here: 

3333333333333333333333333333222222222222
3itemA     33itemB33itemC  3           2
3          33     33       3           2
3          33     33       3           2
3          33     33       3           2
3          33     33       3           2
3          33     33       3           2
3          33     33       3           2
3          33     33       3           2
3333333333333333333333333333222222222222
3333333333333333333333333333333333333222
3itemD      33itemE33itemF 33itemG  3  2
3           33     33      33       3  2
3           33     33      33       3  2
3           33     33      33       3  2
3           33     33      33       3  2
3           33     33      33       3  2
3           33     33      33       3  2
3           33     33      33       3  2
3333333333333333333333333333333333333222

152 chars were rendered offscreen and were skipped
`1` at {X:0 Y:20}
`1` at {X:0 Y:21}
`1` at {X:0 Y:22}
`1` at {X:0 Y:23}
`1` at {X:1 Y:23}
`1` at {X:2 Y:23}
`1` at {X:3 Y:23}
`1` at {X:4 Y:23}
`1` at {X:5 Y:23}
`1` at {X:6 Y:23}
`1` at {X:7 Y:23}
`1` at {X:8 Y:23}
`1` at {X:9 Y:23}
`1` at {X:10 Y:23}
`1` at {X:11 Y:23}
`1` at {X:12 Y:23}
`1` at {X:13 Y:23}
`1` at {X:14 Y:23}
`1` at {X:15 Y:23}
`1` at {X:16 Y:23}
`1` at {X:17 Y:23}
`1` at {X:18 Y:23}
`1` at {X:19 Y:23}
`1` at {X:20 Y:23}
`1` at {X:21 Y:23}
`1` at {X:22 Y:23}
`1` at {X:23 Y:23}
`1` at {X:24 Y:23}
`1` at {X:25 Y:23}
`1` at {X:26 Y:23}
`1` at {X:27 Y:23}
`1` at {X:28 Y:23}
`1` at {X:29 Y:23}
`1` at {X:30 Y:23}
`1` at {X:31 Y:23}
`1` at {X:32 Y:23}
`1` at {X:33 Y:23}
`1` at {X:34 Y:23}
`1` at {X:35 Y:23}
`1` at {X:36 Y:23}
`1` at {X:37 Y:23}
`1` at {X:38 Y:23}
`1` at {X:39 Y:20}
`1` at {X:39 Y:21}
`1` at {X:39 Y:22}
`1` at {X:39 Y:23}
`2` at {X:0 Y:20}
`2` at {X:0 Y:21}
`2` at {X:0 Y:22}
`2` at {X:0 Y:23}
`2` at {X:1 Y:20}
`2` at {X:1 Y:23}
`2` at {X:2 Y:20}
`2` at {X:2 Y:23}
`2` at {X:3 Y:20}
`2` at {X:3 Y:23}
`2` at {X:4 Y:20}
`2` at {X:4 Y:23}
`2` at {X:5 Y:20}
`2` at {X:5 Y:23}
`2` at {X:6 Y:20}
`2` at {X:6 Y:23}
`2` at {X:7 Y:20}
`2` at {X:7 Y:23}
`2` at {X:8 Y:20}
`2` at {X:8 Y:23}
`2` at {X:9 Y:20}
`2` at {X:9 Y:23}
`2` at {X:10 Y:20}
`2` at {X:10 Y:23}
`2` at {X:11 Y:20}
`2` at {X:11 Y:23}
`2` at {X:12 Y:20}
`2` at {X:12 Y:23}
`2` at {X:13 Y:20}
`2` at {X:13 Y:23}
`2` at {X:14 Y:20}
`2` at {X:14 Y:23}
`2` at {X:15 Y:20}
`2` at {X:15 Y:23}
`2` at {X:16 Y:20}
`2` at {X:16 Y:23}
`2` at {X:17 Y:20}
`2` at {X:17 Y:23}
`2` at {X:18 Y:20}
`2` at {X:18 Y:23}
`2` at {X:19 Y:20}
`2` at {X:19 Y:23}
`2` at {X:20 Y:20}
`2` at {X:20 Y:23}
`2` at {X:21 Y:20}
`2` at {X:21 Y:23}
`2` at {X:22 Y:20}
`2` at {X:22 Y:23}
`2` at {X:23 Y:20}
`2` at {X:23 Y:23}
`2` at {X:24 Y:20}
`2` at {X:24 Y:23}
`2` at {X:25 Y:20}
`2` at {X:25 Y:23}
`2` at {X:26 Y:20}
`2` at {X:26 Y:23}
`2` at {X:27 Y:20}
`2` at {X:27 Y:23}
`2` at {X:28 Y:20}
`2` at {X:28 Y:23}
`2` at {X:29 Y:20}
`2` at {X:29 Y:23}
`2` at {X:30 Y:20}
`2` at {X:30 Y:23}
`2` at {X:31 Y:20}
`2` at {X:31 Y:23}
`2` at {X:32 Y:20}
`2` at {X:32 Y:23}
`2` at {X:33 Y:20}
`2` at {X:33 Y:23}
`2` at {X:34 Y:20}
`2` at {X:34 Y:23}
`2` at {X:35 Y:20}
`2` at {X:35 Y:23}
`2` at {X:36 Y:20}
`2` at {X:36 Y:23}
`2` at {X:37 Y:20}
`2` at {X:37 Y:23}
`2` at {X:38 Y:20}
`2` at {X:38 Y:23}
`2` at {X:39 Y:20}
`2` at {X:39 Y:21}
`2` at {X:39 Y:22}
`2` at {X:39 Y:23}
`r` at {X:1 Y:21}
`o` at {X:2 Y:21}
`w` at {X:3 Y:21}
` ` at {X:4 Y:21}
`2` at {X:5 Y:21}
`3` at {X:0 Y:20}
`3` at {X:0 Y:21}
`3` at {X:0 Y:22}
`3` at {X:0 Y:23}
`3` at {X:1 Y:20}
`3` at {X:1 Y:23}
`3` at {X:2 Y:20}
`3` at {X:2 Y:23}
`3` at {X:3 Y:20}
`3` at {X:3 Y:21}
`3` at {X:3 Y:22}
`3` at {X:3 Y:23}
`i` at {X:1 Y:21}
`t` at {X:2 Y:21}
`e` at {X:3 Y:21}
`m` at {X:4 Y:21}
`H` at {X:5 Y:21}

